### PR TITLE
Support version for proposed APIs

### PR DESCRIPTION
**Pull request by sandy081**
Ref: https://github.com/microsoft/vscode/issues/214294

- Support version for API proposals. Version is an incremental number and should be provided in the proposal d.ts file as a comment like below.

```
// version: 1
```

- Updated `generateApiProposalNames` to generate proposals with versions. Moved the `extensionsApiProposals` files to platform so that proposals can be accessed by shared process.

- Extensions can lock the proposal version by mentioning the version while declaring the proposal in `enabledApiProposals` section. Locking the version should make the extension installable or activated in VS Code with the proposal with that version number only. This PR does not have this installation or activation validation which will be implemented in next PR.

``` json
"enabledApiProposals": [
    "testObserver@1",
]
```

- Introduced `enabledApiProposalNames` property in `IExtensionDescription` that contains parsed `enabledApiProposals`. This should be used by `ExtensionsProposedAPI` for validation and modification.

### CODE REVIEW

**Comment by sandy081** [comment]
Seems, there is something broken with my changes. I will take a look and fix.

**Comment by sandy081** [comment]
Fixed!

File: src/vs/workbench/services/extensions/common/extensionsRegistry.ts [hunk 1]
```diff
@@ -652,7 +652,7 @@ schemaRegistry.registerSchema(productSchemaId, {
 					items: {
 						type: 'string',
 						enum: Object.keys(allApiProposals),
-						markdownEnumDescriptions: Object.values(allApiProposals)
+						markdownEnumDescriptions: Object.values(allApiProposals).map(value => value.proposal)
```
Comment (jrieken) [inline src/vs/workbench/services/extensions/common/extensionsRegistry.ts:655]: We should also add the latest version of the proposal to the API, e.g have `fooProposal` and `fooProposal@3`. This ensures extensions with `fooProposal@1` get a squiggle

File: src/vs/workbench/services/extensions/common/extensionsProposedApi.ts [hunk 2]
```diff
@@ -54,16 +54,13 @@ export class ExtensionsProposedApi {
...
-		type Writeable<T> = { -readonly [P in keyof T]: Writeable<T[P]> };
-		const extension = <Writeable<IExtensionDescription>>_extension;
-		const key = ExtensionIdentifier.toKey(_extension.identifier);
+		const key = ExtensionIdentifier.toKey(extension.identifier);
 
 		// warn about invalid proposal and remove them from the list
-		if (isNonEmptyArray(extension.enabledApiProposals)) {
-			extension.enabledApiProposals = extension.enabledApiProposals.filter(name => {
+		if (isNonEmptyArray(extension.enabledApiProposalNames)) {
+			extension.enabledApiProposalNames = extension.enabledApiProposalNames.filter(name => {
```
Comment (jrieken) [inline src/vs/workbench/services/extensions/common/extensionsProposedApi.ts:63]: We now have two lists of enabled proposals and they now diverge. Only the latter `enabledApiProposalNames` is corrected with product-json information and the former is left as authored (untrusted). We should not have this, I would recommended to stick to `enabledApiProposals` and make consumers do the parsing

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/workbench/services/extensions/common/extensionsRegistry.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/workbench/services/extensions/common/extensionsRegistry.ts:655]: This is in `product.json` where we do not want to add versions. I added `version` to the `package.json` schema here already - 

https://github.com/microsoft/vscode/blob/e3e51b079f52f4262a9d3698ccf1ad8ba14335a4/src/vs/workbench/services/extensions/common/extensionsRegistry.ts#L245

**Review by jrieken — APPROVED** [review]

**Pull request merged**